### PR TITLE
Ret 2185 2

### PR DESCRIPTION
**Pull request by robin-ieong**
https://tools.hmcts.net/jira/browse/RET-2185

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x ] No
```

### CODE REVIEW

**Review by JackReevies — APPROVED** [review]

**Review by nickwest1 — CHANGES_REQUESTED** [review]

File: src/main/java/uk/gov/hmcts/ethos/replacement/docmosis/service/EmailService.java [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+                templateId,
+                emailAddress,
+                personalisation,
+                referenceId
+            );
+            log.info("Sending email success. Reference ID: {}", referenceId);
+
+        } catch (NotificationClientException e) {
+            log.warn("Failed to send email. Reference ID: {}. Reason:", referenceId, e);
+        }
```
[the remark this answers is not in the document]
Comment (robin-ieong) [reply src/main/java/uk/gov/hmcts/ethos/replacement/docmosis/service/EmailService.java:34]: If I remember correctly Les said in the refinement to ignore that for now but will confirm with him again once he's back. He did say that the email should be validated and hence we have implemented the email validation which will display an error message and ask user to re-enter the email address if it isn't in the right format.

File: src/main/java/uk/gov/hmcts/ethos/replacement/docmosis/config/NotificationsConfig.java [hunk 2]
```diff
@@ -0,0 +1,18 @@
+package uk.gov.hmcts.ethos.replacement.docmosis.config;
+
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+import uk.gov.service.notify.NotificationClient;
+
+@Configuration
+public class NotificationsConfig {
```
[the remark this answers is not in the document]
Comment (robin-ieong) [reply src/main/java/uk/gov/hmcts/ethos/replacement/docmosis/config/NotificationsConfig.java:12]: Yes, added.

File: src/main/java/uk/gov/hmcts/ethos/replacement/docmosis/controllers/CreateReferralController.java [hunk 3]
```diff
@@ -32,8 +38,13 @@
 @SuppressWarnings({"PMD.UnnecessaryAnnotationValueElement"})
 public class CreateReferralController {
 
+    @Value("${referral.template.id}")
+    @NotEmpty
+    public String referralTemplateId;
```
[the remark this answers is not in the document]
Comment (robin-ieong) [reply src/main/java/uk/gov/hmcts/ethos/replacement/docmosis/controllers/CreateReferralController.java]: Yes, refactored.

File: src/main/java/uk/gov/hmcts/ethos/replacement/docmosis/controllers/ReplyToReferralController.java [hunk 4]
```diff
@@ -31,9 +37,15 @@
 @RequiredArgsConstructor
 @SuppressWarnings({"PMD.UnnecessaryAnnotationValueElement"})
 public class ReplyToReferralController {
```
[the remark this answers is not in the document]
Comment (robin-ieong) [reply src/main/java/uk/gov/hmcts/ethos/replacement/docmosis/controllers/ReplyToReferralController.java:39]: Added

File: src/main/java/uk/gov/hmcts/ethos/replacement/docmosis/controllers/ReplyToReferralController.java [hunk 5]
```diff
@@ -125,18 +167,22 @@ public ResponseEntity<CCDCallbackResponse> initHearingDetailsForReplyToReferral(
...
         if (!verifyTokenService.verifyTokenSignature(userToken)) {
             log.error(INVALID_TOKEN, userToken);
             return ResponseEntity.status(FORBIDDEN.value()).build();
         }
 
         CaseData caseData = ccdRequest.getCaseDetails().getCaseData();
         UserDetails userDetails = userService.getUserDetails(userToken);
+
+        emailService.sendEmail(referralTemplateId, caseData.getReplyToEmailAddress(),
+            ReferralHelper.buildPersonalisation(caseData, TRUE.equals(caseData.getIsJudge()), false));
```
[the remark this answers is not in the document]
Comment (robin-ieong) [reply src/main/java/uk/gov/hmcts/ethos/replacement/docmosis/controllers/ReplyToReferralController.java:187]: That is actually a string value.

File: src/main/java/uk/gov/hmcts/ethos/replacement/docmosis/helpers/ReferralHelper.java [hunk 6]
```diff
@@ -354,4 +361,31 @@ public static void setReferralStatusToClosed(CaseData caseData) {
...
+    public static List<String> validateEmail(String email) {
+        List<String> errors = new ArrayList<>();
+        if (!EmailValidator.getInstance().isValid(email)) {
+            errors.add(INVALID_EMAIL_ERROR_MESSAGE);
+        }
+
+        return errors;
+    }
+
+    public static Map<String, String> buildPersonalisation(CaseData caseData, boolean isJudge, boolean isNewReferral) {
```
[the remark this answers is not in the document]
Comment (robin-ieong) [reply src/main/java/uk/gov/hmcts/ethos/replacement/docmosis/helpers/ReferralHelper.java:385]: added

File: src/main/java/uk/gov/hmcts/ethos/replacement/docmosis/service/EmailService.java [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+import java.util.UUID;
+
+@Slf4j
+@Service
+@RequiredArgsConstructor
+public class EmailService {
+
+    private final NotificationClient emailClient;
+
+    public void sendEmail(String templateId, String emailAddress, Map<String, String> personalisation) {
```
[the remark this answers is not in the document]
Comment (robin-ieong) [reply src/main/java/uk/gov/hmcts/ethos/replacement/docmosis/service/EmailService.java:29]: added

File: src/test/java/uk/gov/hmcts/ethos/replacement/docmosis/helpers/ReferralHelperTest.java [hunk 8]
```diff
@@ -333,6 +335,46 @@ void setReferralStatusToClosed() {
...
         assertEquals(ReferralStatus.CLOSED, caseData.getReferralCollection().get(0).getValue().getReferralStatus());
     }
+
+    @Test
+    void validateEmail() {
+        List<String> errors = new ArrayList<>();
+        assertEquals(errors, ReferralHelper.validateEmail("valid.email@example.com"));
+
+        errors = new ArrayList<>(Arrays.asList(INVALID_EMAIL_ERROR_MESSAGE));
+        assertEquals(errors, ReferralHelper.validateEmail("invalid.email.example"));
```
[the remark this answers is not in the document]
Comment (robin-ieong) [reply src/test/java/uk/gov/hmcts/ethos/replacement/docmosis/helpers/ReferralHelperTest.java]: Refactored to use the contains method.

File: src/main/java/uk/gov/hmcts/ethos/replacement/docmosis/service/EmailService.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (robin-ieong) [reply src/main/java/uk/gov/hmcts/ethos/replacement/docmosis/service/EmailService.java]: As discussed I have changed it to throw a RunTimeException when it fails to send email.

**Review by nickwest1 — APPROVED** [review]

**Pull request merged**